AWS Pricing Data

tsein-bc opened the pull request:
### What does this PR do?

Polling pricing information by configured rate codes and pushing to Datadog as custom metrics.

### Motivation

There is no support for AWS pricing information currently available with existing Datadog integrations. My team plans to use, say for example, the cost of an m5.xlarge EC2 instance over time to make infrastructure decisions.

### Review checklist

- [x] PR has a [meaningful title](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title) or PR has the `no-changelog` label attached
- [x] Feature or bugfix has tests
- [x] Git history is clean
- [x] If PR impacts documentation, docs team has been notified or an issue has been opened on the [documentation repo](https://github.com/DataDog/documentation/issues/new)

### Additional Notes

Still a work in progress, but I wanted to get early reviews.

hithwen commented:
Hallo @tsein-bc thans so much for contributing. Before preceding to review this integration I'd like to make sure you agree to become the code owner and maintainer for this integration. You'll need to be listed here: https://github.com/DataDog/integrations-extras/blob/master/.github/CODEOWNERS

hithwen commented:
Hallo, @tsein-bc our engineering team is reviewing this proposal. We will get to you as soon as possible.

tsein-bc commented:
Thanks, please let me know what the engineering team thinks. I haven't had time to address the CI failure, I'm not very familiar with tox, but it looks like it's not picking up one of my test mocks, `pricing_client_stubber`

tsein-bc commented:
@hithwen I fixed the bugs that were causing CI tests to fail 🎉

Let me know what the engineering team thinks of this contribution, we'd really like to have this functionality available in Datadog to help drive our infrastructure decisions.

tsein-bc commented:
@irabinovitch @hithwen @dsahni Hey guys, have you had a chance to take a look at this PR? Anything I can help with? Thank you!

dsahni commented:
@tsein-bc Thank you so much for your patience! We're still in the process of reviewing this and will get back to you within the next couple of days on next steps.

FlorianVeaux commented on aws_pricing/README.md:
```diff
@@ -0,0 +1,55 @@
+
+This check pulls in pricing information [published by AWS][1] to make it easier to measure cost of resource utilization in within Datadog.
+
+<div>Icons made by <a href="https://www.flaticon.com/authors/eucalyp" title="Eucalyp">Eucalyp</a> from <a href="https://www.flaticon.com/" 			    title="Flaticon">www.flaticon.com</a> is licensed by <a href="http://creativecommons.org/licenses/by/3.0/" 			    title="Creative Commons BY 3.0" target="_blank">CC 3.0 BY</a></div>
```
To be confirmed by the docs team but not sure we support html in markdown, I'll ask them to review.

l0k0ms replied:
Plop, I'm not sure why this content is here. it's not related to the integration itself.. it should be removed i think.

tsein-bc replied:
The I con I used for this integration is from the website above, they ask for attribution on use, so I was trying to abide by that license. I'll change it to markdown.

tsein-bc replied:
Alternatively, if you guys know any designers, I'd be open to accepting a new icon for this integration :)

FlorianVeaux commented on aws_pricing/datadog_checks/aws_pricing/aws_pricing.py:
```diff
@@ -0,0 +1,104 @@
+    rate_code_parts = rate_code.split('.')
+    term_code = '.'.join(rate_code_parts[:2])
+
+    term = list(filter(lambda term: term_code in term, terms))[0]
```
nit:
```python
from six.moves import filter
next(filter(lambda term: term_code in term, terms))
```

py2: List was created twice
py3: List was created once
Now the code just picks the first element which matches the lambda.

FlorianVeaux commented on aws_pricing/datadog_checks/aws_pricing/aws_pricing.py:
```diff
@@ -0,0 +1,104 @@
+
+
+def get_price_from_price_dimensions(price_dimensions):
+    return float(price_dimensions['pricePerUnit']['USD'])
```
Is this consistently in USD?

tsein-bc replied:
From the documentation I've seen -- yes

l0k0ms commented on aws_pricing/README.md:
```diff
@@ -0,0 +1,55 @@
+
+### Installation
+
+The AWS Pricing check is not included in the [Datadog Agent][2] package, so you will
```
Could you please inline [the official instructions to install an integrations from integrations-extras](https://docs.datadoghq.com/agent/guide/community-integrations-installation-with-docker-agent/) ?  Or at least link to it :)

l0k0ms commented on aws_pricing/README.md:
```diff
@@ -0,0 +1,55 @@
+[3]: https://docs.datadoghq.com/agent/faq/agent-commands/#start-stop-restart-the-agent
+[4]: https://docs.datadoghq.com/agent/faq/agent-commands/#agent-status-and-information
+[5]: https://docs.datadoghq.com/help
+[6]: ttps://github.com/DataDog/integrations-extras/blob/master/aws_pricing/metadata.csv
```
```suggestion
[6]: https://github.com/DataDog/integrations-extras/blob/master/aws_pricing/metadata.csv
```

l0k0ms commented on aws_pricing/README.md:
```diff
@@ -0,0 +1,55 @@
+
+[1]: https://aws.amazon.com/pricing/
+[2]: https://github.com/DataDog/integrations-core/blob/master/aws_pricing/datadog_checks/aws_pricing/data/conf.yaml.example
+[3]: https://docs.datadoghq.com/agent/faq/agent-commands/#start-stop-restart-the-agent
```
```suggestion
[3]: https://docs.datadoghq.com/agent/guide/agent-commands/#restart-the-agent
```

l0k0ms commented on aws_pricing/README.md:
```diff
@@ -0,0 +1,55 @@
+
+### Metrics
+
+See [metadata.csv][6] for a list of metrics provided by this check..
```
```suggestion
See [metadata.csv][6] for a list of metrics provided by this check.
```

l0k0ms commented on aws_pricing/README.md:
```diff
@@ -0,0 +1,55 @@
+[1]: https://aws.amazon.com/pricing/
+[2]: https://github.com/DataDog/integrations-core/blob/master/aws_pricing/datadog_checks/aws_pricing/data/conf.yaml.example
+[3]: https://docs.datadoghq.com/agent/faq/agent-commands/#start-stop-restart-the-agent
+[4]: https://docs.datadoghq.com/agent/faq/agent-commands/#agent-status-and-information
```
```suggestion
[4]: https://docs.datadoghq.com/agent/guide/agent-commands/#agent-information
```

l0k0ms commented on aws_pricing/README.md:
```diff
@@ -0,0 +1,55 @@
+
+`aws_pricing.status`:
+
+Returns `Critical` if the Agent encounters an error when using the Boto3 pricing client to collect metrics.
```
```suggestion
Returns `CRITICAL` if the Agent encounters an error when using the Boto3 pricing client to collect metrics.
```

l0k0ms commented on aws_pricing/README.md:
```diff
@@ -0,0 +1,55 @@
+
+Returns `Critical` if the Agent encounters an error when using the Boto3 pricing client to collect metrics.
+
+Returns `Warning` if a rate code was defined in `aws_pricing.d/conf.yaml` which couldn't be found using the Boto3 pricing client.
```
```suggestion
Returns `WARNING` if a rate code was defined in `aws_pricing.d/conf.yaml` which couldn't be found using the Boto3 pricing client.
```

l0k0ms commented on aws_pricing/datadog_checks/aws_pricing/data/conf.yaml.example:
```diff
@@ -0,0 +1,9 @@
+init_config:
+
+instances:
+    ## @param AmazonEC2 - string - optional
```
If the parameter is optional then it should be commented by default, note that you can define a blank instance:

```

instances:
  - 

```

l0k0ms commented on aws_pricing/datadog_checks/aws_pricing/data/conf.yaml.example:
```diff
@@ -0,0 +1,9 @@
+    ## @param AmazonEC2 - string - optional
+    ## Rate codes to monitor for AmazonEC2 service.
+    #
+  - AmazonEC2: rate_code_1,rate_code_2,rate_code_3
```
```suggestion
  - AmazonEC2: <RATE_CODE_1>,<RATE_CODE_2>,<RATE_CODE_3>
```

l0k0ms commented on aws_pricing/datadog_checks/aws_pricing/data/conf.yaml.example:
```diff
@@ -0,0 +1,9 @@
+    ## Rate codes to monitor for AmazonEC2 service.
+    #
+  - AmazonEC2: rate_code_1,rate_code_2,rate_code_3
+    # - AmazonCloudFront: rate_code_1,rate_code_2,rate_code_3
```
I think this can be removed, since we want to give only one example of instance definition.

tsein-bc commented:
@FlorianVeaux @l0k0ms Thanks for the reviews! I don't have much experience with Python, and writing to support both version 2 and 3 threw me for a loop, your advice helped a lot. I think I've addressed all of your comments, please let me know if anything else stands out.

l0k0ms commented on aws_pricing/datadog_checks/aws_pricing/data/conf.yaml.example:
```diff
@@ -0,0 +1,18 @@
+    ## @param region_name - string - required
+    ## AWS region to run PricingService requests in.
+    #
+  - region_name: us-west-2
```
```suggestion
  - region_name: <REGION_NAME>
```

FlorianVeaux commented:
Looks good for me thanks for taking the time to make all those changes. Just one last concern:
Regarding this [comment](https://github.com/DataDog/integrations-extras/pull/414#discussion_r300321285), and this [doc page](https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/using-pelong.html). It looks like only two region are valid, `us-east-1` and `ap-south-1`. I'd like if we can first:
- check if boto works with other endpoints (like `us-west-2`) 
- understand how the result differ between endpoints

Depending on the previous result we might want to have the `region_name` parameter from the config file optional and default to `us-east-1`.

tsein-bc commented:
@FlorianVeaux @l0k0ms Alright, I've applied the last set of change requests

l0k0ms commented on aws_pricing/README.md:
```diff
@@ -0,0 +1,56 @@
+
+## Overview
+
+This check pulls in pricing information [published by AWS][1] to make it easier to measure cost of resource utilization in within Datadog.
```
```suggestion
This check pulls pricing information [published by AWS][1] to make it easier to measure cost of resource utilization within Datadog.
```

l0k0ms commented on aws_pricing/README.md:
```diff
@@ -0,0 +1,56 @@
+
+### Installation
+
+The AWS Pricing check is not included in the [Datadog Agent][2] package, so you will
```
```suggestion
The AWS Pricing check is not included in the [Datadog Agent][2] package, so you
```

FlorianVeaux commented on aws_pricing/datadog_checks/aws_pricing/data/conf.yaml.example:
```diff
@@ -0,0 +1,22 @@
+instances:
+  -
+
+    ## @param region_name - string - optional
```
```suggestion
    ## @param region_name - string - optional - default: us-east-1
```

l0k0ms replied:
we should add it as placeholder also then

FlorianVeaux replied:
You're right 👍

l0k0ms commented on aws_pricing/datadog_checks/aws_pricing/data/conf.yaml.example:
```diff
@@ -0,0 +1,22 @@
+instances:
+  -
+
+    ## @param region_name - string - optional
```
Is there a default value for this parameter ?

FlorianVeaux commented on aws_pricing/datadog_checks/aws_pricing/data/conf.yaml.example:
```diff
@@ -0,0 +1,22 @@
+    ## As of writing, only us-east-1 and ap-south-1 are supported.
+    ## See https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/using-pelong.html
+    #
+    # - region_name: <REGION>
```
```suggestion
    # - region_name: us-east-1
```

tsein-bc commented:
@FlorianVeaux @l0k0ms  I think I got the last of the PR feedback addressed, let me know if you need anything else. Thanks again!

FlorianVeaux commented:
Sorry for the delay here and thanks again 🚀
